src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs

```diff
@@ -525,6 +525,135 @@ public override BoundNode VisitTypeOrInstanceInitializers(BoundTypeOrInstanceIni
+            // code to pretend there are two indexers that support Index and Range
+            // on arrays.
+
+            var syntax = node.Syntax;
+            var F = _factory;
+            var indexLocal = F.StoreToTemp(
+                VisitExpression(node.Indices[0]),
+                out BoundAssignmentOperator indexAssign);
+            var arrayLocal = F.StoreToTemp(
+                VisitExpression(node.Expression),
+                out BoundAssignmentOperator arrayAssign);
+            var indexType = VisitType(node.Indices[0].Type);
+
+            var indexValueSymbol = (PropertySymbol)F.WellKnownMember(WellKnownMember.System_Index__Value);
+            var indexFromEndSymbol = (PropertySymbol)F.WellKnownMember(WellKnownMember.System_Index__FromEnd);
+
+            BoundExpression resultExpr;
+            if (indexType == _compilation.GetWellKnownType(WellKnownType.System_Index))
+            {
+
+                // array[Index] is translated to:
+                // index.FromEnd ? array[array.Length - index.Value] : array[index.Value]
+
+                var indexValueExpr = F.Property(indexLocal, indexValueSymbol);
+
+                resultExpr = F.Sequence(
+                    ImmutableArray.Create<LocalSymbol>(
+                        indexLocal.LocalSymbol,
+                        arrayLocal.LocalSymbol),
+                    ImmutableArray.Create<BoundExpression>(
+                        indexAssign,
+                        arrayAssign),
+                    F.Conditional(
+                        F.Property(indexLocal, indexFromEndSymbol),
+                        F.ArrayAccess(arrayLocal, ImmutableArray.Create<BoundExpression>(F.Binary(
+                            BinaryOperatorKind.Subtraction,
+                            F.SpecialType(SpecialType.System_Int32),
+                            F.ArrayLength(arrayLocal),
+                            indexValueExpr))),
+                        F.ArrayAccess(arrayLocal, ImmutableArray.Create(indexValueExpr)),
+                        node.Type));
+            }
+            else if (indexType == _compilation.GetWellKnownType(WellKnownType.System_Range))
+            {
+                // array[Range] is translated to:
+                // var start = range.Start.FromEnd ? array.Length - range.Start.Value : range.Start.Value;
+                // var end = range.End.FromEnd ? array.Length - range.End.Value : range.End.Value;
+                // var length = end - start;
+                // var newArr = new T[length];
+                // Array.Copy(array, start, newArr, 0, length);
+                // push newArray
+
+                var rangeStartSymbol = (PropertySymbol)F.WellKnownMember(WellKnownMember.System_Range__Start);
+                var rangeEndSymbol = (PropertySymbol)F.WellKnownMember(WellKnownMember.System_Range__End);
+                var arrayCopySymbol = F.WellKnownMethod(WellKnownMember.System_Array__Copy);
```

Comment:
Same comment on the  members being missing

Reply:
Same deal; helper throws on missing members.

Reply:
Ah okay.